README.md
RSocrata
========

**Master** 

[![Linux build - Master](https://img.shields.io/travis/Chicago/RSocrata/master.svg?style=flat-square&label=Linux build)](https://travis-ci.org/Chicago/RSocrata)[![Windows build - Master](https://img.shields.io/appveyor/ci/tomschenkjr/RSocrata/master.svg?style=flat-square&label=Windows build)](https://ci.appveyor.com/project/tomschenkjr/rsocrata/branch/master)[![Coverage - Master](https://img.shields.io/coveralls/Chicago/RSocrata/master.svg?style=flat-square&label=Coverage - Master)](https://coveralls.io/r/Chicago/RSocrata?branch=master)

**Dev**

[![Linux build - Dev](https://img.shields.io/travis/Chicago/RSocrata/dev.svg?style=flat-square&label=Linux build)](https://travis-ci.org/Chicago/RSocrata)[![Windows build - Dev](https://img.shields.io/appveyor/ci/tomschenkjr/RSocrata/dev.svg?style=flat-square&label=Windows build)](https://ci.appveyor.com/project/tomschenkjr/rsocrata/branch/dev)[![Coverage - Dev](https://img.shields.io/coveralls/Chicago/RSocrata/dev.svg?style=flat-square&label=Coverage - Dev)](https://coveralls.io/r/Chicago/RSocrata?branch=dev)

A tool for downloading Socrata datasets as R data frames
--------------------------------------------------------	

Provided with a URL to a dataset resource published on a [Socrata](http://www.socrata.com) webserver,
or a Socrata [SoDA (Socrata Open Data Application Program Interface) web API](http://dev.socrata.com) query,
or a Socrata "human-friendly" URL, ```read.socrata()```
returns an [R data frame](http://stat.ethz.ch/R-manual/R-devel/library/base/html/data.frame.html).
Converts dates to [POSIX](http://stat.ethz.ch/R-manual/R-devel/library/base/html/DateTimeClasses.html) format.
Supports CSV and JSON download file formats from Socrata.
Manages the throttling of data returned from Socrata and allows users to provide an [application token](http://dev.socrata.com/docs/app-tokens.html).
Supports [SoDA query parameters](http://dev.socrata.com/docs/queries.html) in the URL string for further filtering, sorting, and queries.

Use ```ls.socrata()``` to list all datasets available on a Socrata webserver.

This package uses [`testthat`](http://cran.r-project.org/package=testthat) test coverage.

### Installation

Use `devtools` to install the latest version from Github:

```
library(devtools)
devtools::install_github("Chicago/RSocrata")
```

**OR** 

from [CRAN](http://cran.r-project.org/package=RSocrata):

```
install.packages("RSocrata")
```

**Beware**:

In order to support `GeoJSON` (which is semi-optional), it is necessary to install [geojsonio](https://github.com/ropensci/geojsonio) correctly!
This depends on packages such as `rgdal` & `rgeos` (both on CRAN), which additionally on Linux you will need to install through `apt-get`:

`sudo apt-get install libgdal1-dev libgdal-dev libgeos-c1 libproj-dev`

Then install both CRAN packages using:

```
install.packages("rgdal")
install.packages("rgeos")
```

### Examples & Chanelog

Look for examples in the [`vignette` folder](https://github.com/Chicago/RSocrata/blob/dev/vignettes/Examples.Rmd) and see `NEWS.MD` in the root of this repository. 

### Issues

**Please report issues**, request enhancements or fork us at the [City of Chicago github](https://github.com/Chicago/RSocrata/issues).

### Contributing

If you would like to contribute to this project, please see the [contributing documentation](CONTRIBUTING.md) and the [product roadmap](https://github.com/Chicago/RSocrata/wiki/Roadmap#planned-releases).


NEWS
### 1.1 
Add check for valid Socrata resource URL. Add check for supported download file format. Add support for Socrata short dates.

### 1.2 
Use comma-separated file format for Socrata downloads.

### 1.3 
* Added support for human-readable URL. Users can now copy and paste URLs of Socrata-hosted datasets, which will be transformed into a valid SoDA API web query. 

* Added additional RUnit tests to validate new functionality.

### 1.4 
Add json file format for Socrata downloads. Switch to `RJSONIO` from `rjson`. 

### 1.5 Several changes:

* Swapped to ```jsonlite``` from ```RJSONIO```
* Added handling for long and short dates
* Added unit test for reading private datasets

### 1.5.1 
Deprecated ```httr::guess_media()``` and implemented ```mime::guess_type()```

### 1.6.0 Several changes:

* New function, ```ls.socrata``` to list all datasets on a Socrata portal.
* New optional argument, ```app_token```, which lets users supply an API token while using ```read.socrata()``` to minimize throttling.
* Repairs a bug where ```read.socrata``` failed when reading in a date with a column, but there are null values in that column.
* Minor changes to the DESCRIPTION documentation to point users to GitHub for issues and provides new contact information.

### 1.6.1 Bug fixes:

* Resolved potential [name collision issue](https://github.com/Chicago/RSocrata/issues/42)
* Cleaned-up documentation with contributor instructions [#23](https://github.com/Chicago/RSocrata/issues/23) and [#28](https://github.com/Chicago/RSocrata/issues/28))
* Moved test coverage in `RUnit` to `testthat` and implemented code coverage monitoring ([#41](https://github.com/Chicago/RSocrata/issues/41))
* Clean-up DESCRIPTION ([#40](https://github.com/Chicago/RSocrata/issues/40))
* Add continuous integration for Windows ([#39](https://github.com/Chicago/RSocrata/issues/39))
* Migrate Travis-CI to "proper" R YAML ([#46](https://github.com/Chicago/RSocrata/issues/46))


### 1.7 (see roadmap)

* Add a small vignette with existing & new GeoJSON examples with ```leaflet``` map package
* Mostly internal changes which should not influence the current behaviour ([#53](https://github.com/Chicago/RSocrata/pull/53))
* Add support of a `floating timestamp`
* New error handling function
* New functions returning metadata and GeoJSON data (similar to the read.socrata)
* Should be fixing #27 + #24



R/tests/testRSocrata.R
# RUnit tests
# 
# resource 4334-bgaj on the Socrata demo site is USGS Earthquakes for 2012-11-01 API School Demo
#
# Author: Hugh 2013-07-15
###############################################################################

library('RUnit')

source("R/RSocrata.R")

test.posixifyLong <- function() {
	dt <- posixify("09/14/2012 10:38:01 PM")
	checkEquals("POSIXlt", class(dt)[1], "first data type of a date")
	checkEquals(2012, dt$year + 1900, "year")
	checkEquals(9, dt$mon + 1, "month")
	checkEquals(14, dt$mday, "day")
	checkEquals(22, dt$hour, "hours")
	checkEquals(38, dt$min, "minutes")
	checkEquals(1, dt$sec, "seconds")
}

test.posixifyShort <- function() {
	dt <- posixify("09/14/2012")
	checkEquals("POSIXlt", class(dt)[1], "first data type of a date")
	checkEquals(2012, dt$year + 1900, "year")
	checkEquals(9, dt$mon + 1, "month")
	checkEquals(14, dt$mday, "day")
	checkEquals(0, dt$hour, "hours")
	checkEquals(0, dt$min, "minutes")
	checkEquals(0, dt$sec, "seconds")
}

test.readSocrataCsv <- function() {
	df <- read.socrata('https://soda.demo.socrata.com/resource/4334-bgaj.csv')
	checkEquals(1007, nrow(df), "rows")
	checkEquals(9, ncol(df), "columns")
}

test.readSocrataJson <- function() {
	df <- read.socrata('https://soda.demo.socrata.com/resource/4334-bgaj.json')
	checkEquals(1007, nrow(df), "rows")
	checkEquals(11, ncol(df), "columns")
}

test.readSocrataNoScheme <- function() {
	checkException(read.socrata('soda.demo.socrata.com/resource/4334-bgaj.csv'))
}

test.readSoQL <- function() {
	df <- read.socrata('http://soda.demo.socrata.com/resource/4334-bgaj.csv?$select=region')
	checkEquals(1007, nrow(df), "rows")
	checkEquals(1, ncol(df), "columns")
}

test.readSoQLColumnNotFound <- function() {
	# SoQL API uses field names, not human names
	checkException(read.socrata('http://soda.demo.socrata.com/resource/4334-bgaj.csv?$select=Region'))
}

test.readPrivate <- function() {
  checkException(read.socrata('http://data.cityofchicago.org/resource/j8vp-2qpg.json'))
}

test.readSocrataHumanReadable <- function() {
	df <- read.socrata('https://soda.demo.socrata.com/dataset/USGS-Earthquake-Reports/4334-bgaj')
	checkEquals(1007, nrow(df), "rows")
	checkEquals(9, ncol(df), "columns")
}

test.readSocrataFormatNotSupported <- function() {
	# Unsupported data formats
	checkException(read.socrata('http://soda.demo.socrata.com/resource/4334-bgaj.xml'))
}

test.readSocrataCalendarDateLong <- function() {
	df <- read.socrata('http://soda.demo.socrata.com/resource/4334-bgaj.csv')
	dt <- df$Datetime[1] # "2012-09-14 22:38:01"
	checkEquals("POSIXlt", class(dt)[1], "data type of a date")
	checkEquals(2012, dt$year + 1900, "year")
	checkEquals(9, dt$mon + 1, "month")
	checkEquals(14, dt$mday, "day")
	checkEquals(22, dt$hour, "hours")
	checkEquals(38, dt$min, "minutes")
	checkEquals(1, dt$sec, "seconds")
}

test.readSocrataCalendarDateShort <- function() {
	df <- read.socrata('http://data.cityofchicago.org/resource/y93d-d9e3.csv?$order=debarment_date')
	dt <- df$DEBARMENT.DATE[1] # "05/21/1981"
	checkEquals("POSIXlt", class(dt)[1], "data type of a date")
	checkEquals(81, dt$year, "year")
	checkEquals(5, dt$mon + 1, "month")
	checkEquals(21, dt$mday, "day")
	checkEquals(0, dt$hour, "hours")
	checkEquals(0, dt$min, "minutes")
	checkEquals(0, dt$sec, "seconds")
}

test.readSocrataNADatesCsv {
  # df <- read.socrata('http://data.cityofboston.gov/resource/awu8-dc52.csv?property_id=177808') # This was the original query which produced the error
  ## The above statement is no longer valid because property_id 177808 is no longer online.
  df <- read.socrata('http://data.cityofboston.gov/resource/awu8-dc52.csv') # Seems to fail with https, takes forever
}

test.isFourByFour <- function() {
	checkTrue(isFourByFour("4334-bgaj"), "ok")
	checkTrue(!isFourByFour("4334c-bgajc"), "11 characters instead of 9")
	checkTrue(!isFourByFour("433-bga"), "7 characters instead of 9")
	checkTrue(!isFourByFour("433-bgaj"), "3 characters before dash instead of 4")
	checkTrue(!isFourByFour("4334-!gaj"), "non-alphanumeric character")
}

test.isFourByFourUrl <- function() {
	checkException(read.socrata("https://soda.demo.socrata.com/api/views/4334c-bgajc"), "11 characters instead of 9")
	checkException(read.socrata("https://soda.demo.socrata.com/api/views/433-bga"), "7 characters instead of 9")
	checkException(read.socrata("https://soda.demo.socrata.com/api/views/433-bgaj"), "3 characters before dash instead of 4")
	checkException(read.socrata("https://soda.demo.socrata.com/api/views/4334-!gaj"), "non-alphanumeric character")
}

test.readSocrataInvalidUrl <- function() {
	checkException(read.socrata("a.fake.url.being.tested"), "invalid url")
}

test.readSocrataToken <- function(){
  df <- read.socrata('https://soda.demo.socrata.com/resource/4334-bgaj.csv', app_token="ew2rEMuESuzWPqMkyPfOSGJgE")
  checkEquals(1007, nrow(df), "rows")
  checkEquals(9, ncol(df), "columns")  
}

test.readSocrataHumanReadableToken <- function(){
  df <- read.socrata('https://soda.demo.socrata.com/dataset/USGS-Earthquake-Reports/4334-bgaj', app_token="ew2rEMuESuzWPqMkyPfOSGJgE")
  checkEquals(1007, nrow(df), "rows")
  checkEquals(9, ncol(df), "columns")  
}

test.readAPIConflict <- function(){
  df <- read.socrata('https://soda.demo.socrata.com/resource/4334-bgaj.csv?$$app_token=ew2rEMuESuzWPqMkyPfOSGJgE', app_token="ew2rEMuESuzWPqMkyPfOSUSER")
  checkEquals(1007, nrow(df), "rows")
  checkEquals(9, ncol(df), "columns")
  # Check that function is calling the API token specified in url
  checkTrue(substr(validateUrl('https://soda.demo.socrata.com/resource/4334-bgaj.csv?$$app_token=ew2rEMuESuzWPqMkyPfOSGJgE', app_token="ew2rEMuESuzWPqMkyPfOSUSER"), 70, 94)=="ew2rEMuESuzWPqMkyPfOSGJgE")
}

test.readAPIConflictHumanReadable <- function(){
  df <- read.socrata('https://soda.demo.socrata.com/dataset/USGS-Earthquake-Reports/4334-bgaj?$$app_token=ew2rEMuESuzWPqMkyPfOSGJgE', app_token="ew2rEMuESuzWPqMkyPfOSUSER")
  checkEquals(1007, nrow(df), "rows")
  checkEquals(9, ncol(df), "columns")
  # Check that function is calling the API token specified in url
  checkTrue(substr(validateUrl('https://soda.demo.socrata.com/dataset/USGS-Earthquake-Reports/4334-bgaj?$$app_token=ew2rEMuESuzWPqMkyPfOSGJgE', app_token="ew2rEMuESuzWPqMkyPfOSUSER"), 70, 94)=="ew2rEMuESuzWPqMkyPfOSGJgE")
}

test.incorrectAPIQuery <- function(){
  # The query below is missing a $ before app_token.
  checkException(read.socrata("https://soda.demo.socrata.com/resource/4334-bgaj.csv?$app_token=ew2rEMuESuzWPqMkyPfOSGJgE"))
  # Check that it was only because of missing $  
  df <- read.socrata("https://soda.demo.socrata.com/resource/4334-bgaj.csv?$$app_token=ew2rEMuESuzWPqMkyPfOSGJgE")
  checkEquals(1007, nrow(df), "rows")
  checkEquals(9, ncol(df), "columns") 
}

test.incorrectAPIQueryHumanReadable <- function(){
  # The query below is missing a $ before app_token.
  checkException(read.socrata("https://soda.demo.socrata.com/dataset/USGS-Earthquake-Reports/4334-bgaj?$app_token=ew2rEMuESuzWPqMkyPfOSGJgE"))
  # Check that it was only because of missing $  
  df <- read.socrata("https://soda.demo.socrata.com/dataset/USGS-Earthquake-Reports/4334-bgaj?$$app_token=ew2rEMuESuzWPqMkyPfOSGJgE")
  checkEquals(1007, nrow(df), "rows")
  checkEquals(9, ncol(df), "columns") 
}

test.lsSocrata <- function() {
    # Makes some potentially erroneous assumptions about availability
    # of soda.demo.socrata.com
    df <- ls.socrata("https://soda.demo.socrata.com")
    checkEquals(TRUE, nrow(df) > 0)
    # Test comparing columns against data.json specifications:
    # https://project-open-data.cio.gov/v1.1/schema/
    core_names <- as.character(c("issued", "modified", "keyword", "landingPage", "theme", 
                                 "title", "accessLevel", "distribution", "description", 
                                 "identifier", "publisher", "contactPoint", "license"))
    checkEquals(as.logical(rep(TRUE, length(core_names))), core_names %in% names(df))
    # Check that all names in data.json are accounted for in ls.socrata return
    checkEquals(as.logical(rep(TRUE, length(names(df)))), names(df) %in% c(core_names))
}

test.lsSocrataInvalidURL <- function() {
    checkException(read.socrata("a.fake.url.being.tested"), "invalid url")
}

test.suite <- defineTestSuite("test Socrata SODA interface",
                              dirs = file.path("R/tests"),
                              testFileRegexp = '^test.*\\.R')

runAllTests <- function() { # Run during development, will complete regardless of errors
	test.result <- runTestSuite(test.suite)
	printTextProtocol(test.result) 
}

runAllTestsCI <- function() { # Ran for continuous integration tests, will stop if error found
  test.result <- runTestSuite(test.suite)
  if(getErrors(test.result)$nErr > 0 | getErrors(test.result)$nFail > 0) stop("TEST HAD ERRORS!")
}

tests/prepareUnitTests.R
library("RUnit")

runAllTests <- function(){
    if(length(grep("Rcheck", getwd())) > 0){
      path = file.path(getwd(), "unitTests")
      pkgPath <- file.path(getwd(), "..", "00_pkg_src", "RSocrata", "R")
    } else {
      if(length(grep("tests", getwd())) > 0){
        path = "unitTests"
        pkgPath <- file.path(getwd(), "..", "R")
      } else {
        path = "tests/unitTests"
        pkgPath <- file.path(getwd(), "R")
      }
    }
    source(file.path(pkgPath, "RSocrata.R"))
        
    test.suite <- defineTestSuite("test Socrata SODA interface",
                                  dirs = file.path(path),
                                  testFileRegexp = 'testRSocrata.R')
        
    test.result <- runTestSuite(test.suite)
    
    test.errors <- getErrors(test.result)
    
    printTextProtocol(test.result)
    if(test.errors$nErr > 0 | test.errors$nFail >0) stop("TEST HAD ERRORS!")
}

runAllTests()

R/utils.R
#' Checks the validity of the syntax for a potential Socrata dataset Unique Identifier, also known as a 4x4.
#'
#' @description Will check the validity of a potential dataset unique identifier
#' supported by Socrata. It will provide an exception if the syntax
#' does not align to Socrata unique identifiers. It only checks for
#' the validity of the syntax, but does not check if it actually exists.
#' 
#' @param fourByFour - a string; character vector of length one
#' @return TRUE if is valid Socrata unique identifier, FALSE otherwise
#' @author Tom Schenk Jr \email{tom.schenk@@cityofchicago.org}
#' @examples 
#' isFourByFour(fourByFour = "4334-bgaj")
#' isFourByFour("433-bgaj")
#' isFourByFour(fourByFour = "4334-!gaj")
#' @export
isFourByFour <- function(fourByFour = "") {
  
  if (nchar(fourByFour) == 9) {
    if (identical(grepl("[[:alnum:]]{4}-[[:alnum:]]{4}", fourByFour), TRUE)) {
      return(TRUE)
    } else {
      return(FALSE)
    }
  } else {
    return(FALSE)
  }
  
}

# Convert Socrata human-readable column name to field name
# 
# @description Convert Socrata human-readable column name,
# as it might appear in the first row of data,
# to field name as it might appear in the HTTP header;
# that is, lower case, periods replaced with underscores
# 
# @param humanName - a Socrata human-readable column name
# @return Socrata field name in lower case
# @author Hugh J. Devlin, Ph. D. \email{Hugh.Devlin@@cityofchicago.org}
# @examples
# fieldName("Number.of.Stations") # number_of_stations
# @noRd
# @export
fieldName <- function(humanName = "") {
  tolower(gsub('\\.', '_', humanName))
}

# Convert Socrata calendar_date string to POSIX
# 
# @description Datasets will either specify what timezone they should be interpreted in, 
# or you can usually assume they are in the timezone of the publisher. See examples below too. 
# 
# @seealso \url{http://dev.socrata.com/docs/datatypes/floating_timestamp.html}
# @param x - character vector in one of possible Socrata calendar_date formats
# @return a POSIX date
# @author Hugh J. Devlin, Ph. D. \email{Hugh.Devlin@@cityofchicago.org}
# @examples 
# posixify("2014-10-13T23:00:00")
# posixify("09/14/2012 10:38:01 PM")
# posixify("09/14/2012")
# @noRd 
# @export
posixify <- function(x = "") {
  
  # https://github.com/Chicago/RSocrata/issues/24
  # If a query with a date column returns no data (e.g. NA), posixify would fail without this
  if (length(x) == 0) {
    return(x)
  }
  
  # Three calendar date formats supplied by Socrata
  # See https://github.com/GregDThomas/jquery-localtime/issues/1 for the floating timestamp regex
  
  if (any(regexpr("^(-?(?:[1-9][0-9]*)?[0-9]{4})-(1[0-2]|0[1-9])-(3[0-1]|0[1-9]|[1-2][0-9])T(2[0-3]|[0-1][0-9]):([0-5][0-9]):([0-5][0-9])(.[0-9]+)?(Z|[+-](?:2[0-3]|[0-1][0-9]):[0-5][0-9])?$", x)[1] == TRUE)) { 
    # floating timestamp
    strptime(x, format = "%Y-%m-%dT%H:%M:%S") 
    
  } else if (any(regexpr("^[[:digit:]]{1,2}/[[:digit:]]{1,2}/[[:digit:]]{4}$", x)[1] == TRUE)) {
    # short date format
    strptime(x, format = "%m/%d/%Y") 
    
  } else {
    # long date-time format
    strptime(x, format = "%m/%d/%Y %I:%M:%S %p") 
    
  }
  
}

# Clean everything after "?", "&" or "."
# 
# @source https://stackoverflow.com/questions/5631384/remove-everything-after-a-certain-character
# @source http://rfunction.com/archives/1499
# 
# @examples
# cleanQuest(url = "http://data.cityofchicago.org/resource/y93d-d9e3.csv?%24order=debarment_date&%24limit=50000")
# @returns http://data.cityofchicago.org/resource/y93d-d9e3.csv
# @export
cleanQuest <- function(url = "") {
  cleanURL <- strsplit(url, "?",  fixed = TRUE)
  return(cleanURL[[1]][1])
}

# @export
cleanAmp <- function(url = "") {
  cleanURL <- strsplit(url, "&",  fixed = TRUE)
  return(cleanURL[[1]][1])
}

# @export
cleanDot <- function(url = "") {
  cleanURL <- strsplit(url, ".",  fixed = TRUE)
  return(cleanURL[[1]][1])
}



R/valid.R


getQuery <- function(query){
  if(is.list(query)){
    result <- query
  } else if (is.null(query)){
    result <- NULL
  } else {
    if(substr(query, 1,1) != "?") {
      query <- paste0("?", query)
    }
    result <- httr::parse_url(query)$query
  }
  return(result)
}

getResourcePath <- function(resourcePath,  defaultDataType = "csv",  validDataTypes = c("csv", "json")) {
  ## Extract 4x4
  fourByFour <- substr(basename(as.character(resourcePath)), 1, 9)
  ## Test validity of fourByFour
  if(!isFourByFour(fourByFour)){
    stop(paste(fourByFour, "is not a valid Socrata 4x4"))
  }
  ## Extract mime type
  mimeType <- gsub(".+\\.", "", resourcePath)
  ## Test validity of mimeType
  if(!mimeType %in% validDataTypes) {
    mimeType <- defaultDataType
  }
  ## Construct valid resource path
  result <- file.path("resource", paste0(fourByFour, ".", mimeType))
  return(list(resourcePath = result,
              fourByFour = fourByFour,
              mimeType = mimeType))
  
}

validateUrlQuery <- function(urlParsed,apptoken,keyfield) {
  
  ## Insert token into query part of url
  urlParsed[["query"]] <- insertToken(urlParsed = urlParsed, apptoken = apptoken)
  
  ## Extract the key field (aka order) from the parsedUrl, if it exists
  ## And remove $order from query
  orderArg <- grep(pattern = "\\$order", 
                   x = names(urlParsed[["query"]]), 
                   ignore.case = TRUE)
  if(length(orderArg) > 0){
    if(is.null(keyfield)){
      ## Use the order argument for keyfield
      keyfield <- urlParsed[["query"]][[orderArg]]
    } else {
      if(keyfield != urlParsed[["query"]][[orderArg]]) {
        warning(paste0("The keyfield provided does not match the keyfield ",
                       "implied by the URL argument / arguments\n",
                       "KEYFIELD SUPPLIED: ", keyfield, "\n",
                       "ORDER ARGUMENT IN URL: ", 
                       urlParsed[["query"]][[orderArg]], "\n",
                       "Defaulting to ", keyfield))
      }
    }
    ## remove $order from parsed url
    urlParsed[["query"]][[orderArg]] <- NULL
  }
  
  ## Rely on constructed limit argument, so remove any previous limit
  ## specification from query (it's interger(0) if no limit arg present)
  limitArg <- grep(pattern = "\\$limit",  x = names(urlParsed[["query"]]),   ignore.case = TRUE)
  if(length(limitArg) > 0){
    rowLimit <- urlParsed[["query"]][[limitArg]]
    urlParsed[["query"]][[limitArg]] <- NULL
  } else {
    rowLimit <- NULL
  }
  
  ## If there is nothing left in the query then remove it
  if(length(urlParsed[["query"]]) == 0){
    urlParsed[["query"]] <- NULL
  }
  
  return(list(query = urlParsed$query,
              keyfield = keyfield,
              rowLimit = rowLimit))
}


insertToken <- function(urlParsed, apptoken = NULL) {
  
  ## Extract just the query from the URL
  query <- urlParsed[["query"]]
  
  if(!is.null(apptoken)){
    ## If there is no query at all, create a base
    if(is.null(query)){
      query  <- list()
    }
    ## Convert to character, just in case
    apptoken <- as.character(apptoken)
    ## If the user supplied apptoken as an argument, and there was already 
    ## one in the query, and they don't match, then warn
    queryHasAppToken <- !is.null(query["$$app_token"][[1]])
    if(queryHasAppToken && query[["$$app_token"]] != apptoken){
      msg <- paste0("url = ", httr::build_url(urlParsed), "\n",
                    "The supplied url has an API token embedded in ",
                    "it, and the embedded token (",
                    query[['$$app_token']], ") does not match the ",
                    "supplied apptoken (", apptoken ,").\n",
                    "Using apptoken.")
      warning(msg)
    }
    
    ## Insert apptoken into the query (even if it already exists)
    query[["$$app_token"]] <- apptoken
  }
  return(query)
}

R/returnData.R
# An interface to data hosted online in Socrata data repositories
# This is the main file which uses other functions to download data from a Socrata repositories
#
# Author: Hugh J. Devlin, Ph. D. 2013-08-28
###############################################################################

# library("httr")       # for access to the HTTP header
# library("jsonlite")   # for parsing data types from Socrata
# library("mime")       # for guessing mime type
# library("geojsonio")  # for geospatial json

#' Content parsers
#'
#' Return a data frame for csv or json
#'
#' @author Hugh J. Devlin \email{Hugh.Devlin@@cityofchicago.org}
#' @importFrom httr content
#' @importFrom geojsonio geojson_read
#' @param response - an httr response object
#' @return data frame, possibly empty
#' @noRd
getContentAsDataFrame <- function(response) {
  
  mimeType <- response$header$'content-type'
  
  # skip optional parameters
  sep <- regexpr(';', mimeType)[1]
  
  if (sep != -1) {
    mimeType <- substr(mimeType, 0, sep[1] - 1)
  }
  
  switch(mimeType,
         "text/csv" = 
           httr::content(response), # automatic parsing
         "application/json" = 
           if (httr::content(response, as = "text") == "[ ]") { # empty json?
             data.frame() # empty data frame
           } else {
             data.frame(t(sapply(httr::content(response), unlist)), stringsAsFactors = FALSE)
           }
  ) 
  
}


#' Get a full Socrata data set as an R data frame
#'
#' @description Manages throttling and POSIX date-time conversions.
#'
#' @param url - A Socrata resource URL, or a Socrata "human-friendly" URL, 
#' or Socrata Open Data Application Program Interface (SODA) query 
#' requesting a comma-separated download format (.csv suffix), 
#' May include SoQL parameters, and it is now assumed to include SODA \code{limit} 
#' & \code{offset} parameters.
#' Either use a compelete URL or use parameters below to construct your URL. 
#' @param app_token - a (non-required) string; SODA API token can be used to query the data 
#' portal \url{http://dev.socrata.com/consumers/getting-started.html}
#' @param query - Based on query language called the "Socrata Query Language" ("SoQL"), see 
#' \url{http://dev.socrata.com/docs/queries.html}.
#' @param limit - defaults to the max of 50000. See \url{http://dev.socrata.com/docs/paging.html}.
#' @param offset - defaults to 0. See \url{http://dev.socrata.com/docs/paging.html}.
#' @param output - defaults to csv; one of \code{"csv" or "json"}. 
#' @param domain - A Socrata domain, e.g \url{http://data.cityofchicago.org} 
#' @param fourByFour - a unique 4x4 identifier, e.g. "ydr8-5enu". See more \code{\link{isFourByFour}}
#' 
#' @section TODO: \url{https://github.com/Chicago/RSocrata/issues/14}
#' @section Issue: If you get something like \code{Error in rbind(deparse.level, ...) : 
#' numbers of columns of arguments do not match} when using "json" output, this is a known bug 
#' \url{https://github.com/Chicago/RSocrata/issues/19}! Use instead csv output for time being. 
#'
#' @return a data frame with POSIX dates in csv or json format. 
#' @author Hugh J. Devlin, Ph. D. \email{Hugh.Devlin@@cityofchicago.org}
#' 
#' @examples
#' \dontrun{
#' df_csv <- read.socrata(url = "http://soda.demo.socrata.com/resource/4334-bgaj.csv")
#' df_manual2<-read.socrata(domain="http://data.cityofchicago.org/",fourByFour = "ydr8-5enu")
#' df_manual3<-read.socrata(domain="http://data.cityofchicago.org/",
#' fourByFour="ydr8-5enu",output="csv")
#' }
#' 
#' @importFrom httr parse_url build_url
#' @importFrom mime guess_type
#' @importFrom plyr rbind.fill
#' @importFrom dplyr bind_rows
#' @importFrom data.table rbindlist
#' 
#' @export
read.socrata <- function(url = NULL, app_token = NULL, limit = 50000, domain = NULL, fourByFour = NULL, 
                         query = NULL, offset = 0, output = "csv") {
  
  if (is.null(url) == TRUE) {
    buildUrl <- paste0(domain, "resource/", fourByFour, ".", output)
    url <- httr::parse_url(buildUrl)
  }
  
  # check url syntax, allow human-readable Socrata url
  validUrl <- validateUrl(url, app_token) 
  parsedUrl <- httr::parse_url(paste0(validUrl, "&$limit=", limit))
  
  mimeType <- mime::guess_type(cleanAmp(parsedUrl$path))
  
  if (!(mimeType %in% c("text/csv","application/json", "text/plain"))) {
    stop(mimeType, " not a supported data format. Try JSON or CSV. For GeoJSON use: read.socrataGEO")
  }
  
  response <- errorHandling(validUrl)
  results <- getContentAsDataFrame(response)
  dataTypes <- getSodaTypes(response)
  
  rowCount <- as.numeric(getMetadata(cleanQuest(validUrl))[1])
  
  ## More to come? Loop over pages implicitly
  while (nrow(results) < rowCount) { 
    query_url <- paste0(validUrl, ifelse(is.null(parsedUrl$query), "?", "&"), "$offset=", nrow(results), "&$limit=", limit)
    response <- errorHandling(query_url)
    page <- getContentAsDataFrame(response)
    results <- data.table::rbindlist(list(results, page), fill = TRUE) # accumulate data
  }	
  
  # Convert Socrata calendar dates to POSIX format
  # Check for column names that are not NA and which dataType is a "calendar_date". If there are some, 
  # then convert them to POSIX format
  for (columnName in colnames(results)[!is.na(dataTypes[fieldName(colnames(results))]) 
                                       & dataTypes[fieldName(colnames(results))] == "calendar_date"]) {
    results[[columnName]] <- posixify(results[[columnName]])
  }
  
  return(results)
}

#' Download GeoJSON data using geojsonio package
#' 
#' @param what - \link{geojsonio} What to return format is choosen. One of list (default) or \code{sp}.
#' @param parse - \link{geojsonio} Parse geojson to data.frame like structures if possible or not. Default: FALSE (~not)
#' @param method - \link{geojsonio} One of "web" or "local" (default). Matches on partial strings.
#' @param ... - other arguments from \link{geojsonio} package for geojson_read method
#' @param url - A Socrata resource URL, or a Socrata "human-friendly" URL,
#' requesting a .geojson suffix.
#' 
#' @importFrom geojsonio geojson_read
#' @importFrom httr build_url parse_url
#' @importFrom mime guess_type
#' 
#' @return Returns a \code{sp} object, which is the default option here. 
#'
#' @examples 
#' \dontrun{
#' df_geo <- read.socrataGEO(url = "https://data.cityofchicago.org/resource/6zsd-86xi.geojson")
#' }
#' 
#' @export
read.socrataGEO <- function(url = "", method = "local", what = "sp", parse = FALSE, ...) {
  
  validUrl <- httr::parse_url(url)
  mimeType <- mime::guess_type(validUrl$path)
  
  if (mimeType == "application/vnd.geo+json") {
    results <- geojsonio::geojson_read(url, method = method, parse = parse, what = what, ...)
  } 
  
  return(results)
}  

#' Get the SoDA 2 data types
#'
#' Get the Socrata Open Data Application Program Interface data types from the http response header. 
#' Used only for CSV and JSON, not GeoJSON
#' 
#' @author Hugh J. Devlin, Ph. D. \email{Hugh.Devlin@@cityofchicago.org}
#' @param response - headers attribute from an httr response object
#' @return a named vector mapping field names to data types
#' @importFrom jsonlite fromJSON
#' @noRd
getSodaTypes <- function(response) {
  result <- jsonlite::fromJSON(response$headers[['x-soda2-types']])
  names(result) <- jsonlite::fromJSON(response$headers[['x-soda2-fields']])
  return(result)
}

